File: 0039_combinationSum/main.py
from typing import List
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        subset = []
        def dfs(startIdx, target):
            if startIdx >= len(candidates):
                return
            if target < 0:
                return
            if target == 0:
                res.append(subset[:])
                return
            
            subset.append(candidates[startIdx])
            dfs(startIdx, target - candidates[startIdx])

            subset.pop()
            dfs(startIdx + 1, target)
        
        dfs(0, target)
        return res
    
class Solution1:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        def dfs(cur, cur_sum, idx):
            if cur_sum > target:
                return
            if cur_sum == target:
                res.append(cur)
                return
            
            for i in range(idx, len(candidates)):
                dfs(cur + [candidates[i]], cur_sum + candidates[i], i)
            
        dfs([], 0, 0)
        return res
    
class Solution2:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ways_hash = [ [] for index in range(0, target + 1)]
        ways_hash[0] = [[]]
        
        for candidate in candidates:
            for index in range(candidate, len(ways_hash)):
                for combo in ways_hash[index - candidate]:
                    newCombo = combo + [candidate]
                    ways_hash[index].append(newCombo)

        return ways_hash[target]

        

File: 0039_combinationSum/main.cpp
#include <vector>

class Solution {
private:
    void dfs(std::vector<std::vector<int>>& res, std::vector<int>& subset, 
            std::vector<int>& candidates, int startIdx, int target) {
        if(startIdx >= candidates.size() || target < 0)
            return;
        if(target == 0) {
            res.push_back(subset);
            return;
        }

        subset.push_back(candidates[startIdx]);
        dfs(res, subset, candidates, startIdx, target - candidates[startIdx]);

        subset.pop_back();
        dfs(res, subset, candidates, startIdx + 1, target);
    }
public:
    std::vector<std::vector<int>> combinationSum(std::vector<int>& candidates, int target) {
        std::vector<std::vector<int>> res;
        std::vector<int> subset;
        dfs(res, subset, candidates, 0, target);
        return res;
    }
};